File: problems/python/src/integer_to_roman.py
from abc import ABC, abstractmethod
from typing import override

from problems.python.src.util.dicts import floor_key


class Problem(ABC):
    @abstractmethod
    def int_to_roman(self, num: int) -> str:
        pass


class Default(Problem):
    @override
    def int_to_roman(self, num: int) -> str:
        romans = {
            1000: 'M',
            900: 'CM',
            500: 'D',
            400: 'CD',
            100: 'C',
            90: 'XC',
            50: 'L',
            40: 'XL',
            10: 'X',
            9: 'IX',
            5: 'V',
            4: 'IV',
            1: 'I',
        }
        return self.recurse(romans, num)

    def recurse(self, romans: dict[int, str], num: int) -> str:
        floor = floor_key(romans, num)
        if num == floor:
            return romans[num]
        return romans[floor] + self.recurse(romans, num - floor)


SOLUTIONS: list[Problem] = [Default()]

File: problems/src/main/java/IntegerToRoman.java
import java.util.TreeMap;

public enum IntegerToRoman {
    DEFAULT {
        @Override
        public String intToRoman(int num) {
            TreeMap<Integer, String> romans = new TreeMap<>();
            romans.put(1000, "M");
            romans.put(900, "CM");
            romans.put(500, "D");
            romans.put(400, "CD");
            romans.put(100, "C");
            romans.put(90, "XC");
            romans.put(50, "L");
            romans.put(40, "XL");
            romans.put(10, "X");
            romans.put(9, "IX");
            romans.put(5, "V");
            romans.put(4, "IV");
            romans.put(1, "I");
            return recurse(romans, num);
        }

        public String recurse(TreeMap<Integer, String> romans, int num) {
            int floor = romans.floorKey(num);
            if (num == floor) {
                return romans.get(num);
            }
            return romans.get(floor) + recurse(romans, num - floor);
        }
    };

    public abstract String intToRoman(int num);
}
